csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -635,57 +641,69 @@ public void AddDefault(string facet, IObject servant)
+                    // requests or connections. This ensures that once ShutdownAsync returns, no new requests will be
+                    // dispatched. Calling ToArray is important here to ensure that all the ShutdownAsync calls are
+                    // executed before we eventually hit an await (we want to make that once ShutdownAsync returns a
+                    // Task, all the connections started closing).
+                    // Once _shutdownTask is non null, _incomingConnectionfactories cannot change, so no need to lock
+                    // _mutex.
+                    Task[] tasks = _incomingConnectionFactories.Select(factory => factory.ShutdownAsync()).ToArray();
+
+                    // Wait for activation to complete. This is necessary avoid out of order locator updates.
+                    // _activateTask is readonly once _shutdownTask is non null.
+                    if (_activateTask != null)
+                    {
+                        try
+                        {
+                            await _activateTask.ConfigureAwait(false);
+                        }
+                        catch
+                        {
+                            // Ignore
+                        }
+                    }
+
                     try
                     {
-                        await _activateTask.ConfigureAwait(false);
+                        await UnregisterEndpointsAsync(default).ConfigureAwait(false);
                     }
                     catch
                     {
-                        // Ignore
+                        // We can't throw exceptions in deactivate so we ignore failures to unregister endpoints
                     }
-                }
 
-                try
-                {
-                    await UnregisterEndpointsAsync(default).ConfigureAwait(false);
-                }
-                catch
-                {
-                    // We can't throw exceptions in deactivate so we ignore failures to unregister endpoints
-                }
+                    if (_colocatedConnectionFactory != null)
+                    {
+                        await _colocatedConnectionFactory.ShutdownAsync().ConfigureAwait(false);
+                    }
+
+                    // Wait for the incoming connection factories to be shut down.
+                    await Task.WhenAll(tasks).ConfigureAwait(false);
 
-                if (_colocatedConnectionFactory != null)
+                    // TODO jose: Clear the outgoing connections adapter?
+                    Communicator.EraseRouterInfo(_routerInfo?.Router);
+                    Communicator.RemoveObjectAdapter(this);
+                }
+                finally
                 {
-                    await _colocatedConnectionFactory.ShutdownAsync().ConfigureAwait(false);
+                    _shutdownCompleteSource.TrySetResult(null);
```

Comment:
Like for the communicator shutdown, it would be better to ensure the continuation is ran asynchronously in case it tries to call the synchronous `Dispose()` on the OA.

Reply:
Fixed as suggested.